File: Instrument/OSX/InstrumentDemoFramework/InstrumentDemoViewController.m
/*
	Copyright (C) 2016 Apple Inc. All Rights Reserved.
	See LICENSE.txt for this sample’s licensing information
	
	Abstract:
	View controller for the InstrumentDemo audio unit. Manages the interactions between a InstrumentView and the audio unit's parameters.
*/

#import <Cocoa/Cocoa.h>
#import "InstrumentDemoViewController.h"
#import <InstrumentDemoFramework/InstrumentDemo.h>

@interface InstrumentDemoViewController () {
    __weak IBOutlet NSSlider    *attackSlider;
    __weak IBOutlet NSSlider    *releaseSlider;

    __weak IBOutlet NSTextField *attackField;
    __weak IBOutlet NSTextField *releaseField;
    
    AUParameter *attackParameter;
    AUParameter *releaseParameter;
    AUParameterObserverToken parameterObserverToken;
}

-(IBAction) attackValueChanged:(id) sender;
-(IBAction) releaseValueChanged:(id) sender;

@end

@implementation InstrumentDemoViewController

-(void)setAudioUnit:(AUv3InstrumentDemo *)audioUnit {
    _audioUnit = audioUnit;
    dispatch_async(dispatch_get_main_queue(), ^{
        if ([self isViewLoaded]) {
            [self connectViewWithAU];
        }
    });
}

-(AUv3InstrumentDemo *)getAudioUnit {
    return _audioUnit;
}

- (void)dealloc {
    if (parameterObserverToken) {
        [_audioUnit.parameterTree removeParameterObserver:parameterObserverToken];
        parameterObserverToken = 0;
    }
}

-(void)connectViewWithAU {
    AUParameterTree *paramTree = _audioUnit.parameterTree;

    if (paramTree) {
        attackParameter = [paramTree valueForKey: @"attack"];
        releaseParameter = [paramTree valueForKey: @"release"];

        __weak InstrumentDemoViewController *weakSelf = self;
        __weak AUParameter *weakAttackParameter = attackParameter;
        __weak AUParameter *weakReleaseParameter = releaseParameter;
        parameterObserverToken = [paramTree tokenByAddingParameterObserver:^(AUParameterAddress address, AUValue value) {
            __strong InstrumentDemoViewController *strongSelf = weakSelf;
            __strong AUParameter *strongAttackParameter = weakAttackParameter;
            __strong AUParameter *strongReleaseParameter = weakReleaseParameter;
            dispatch_async(dispatch_get_main_queue(), ^{
                if (address == strongAttackParameter.address) {
                    [strongSelf updateAttack];
                } else if (address == strongReleaseParameter.address) {
                    [strongSelf updateRelease];
                }
            });
        }];
        
        [self updateAttack];
        [self updateRelease];
    }
}

- (void)viewDidLoad {
    [super viewDidLoad];

    if (_audioUnit) {
        [self connectViewWithAU];
    }
    
    self.preferredContentSize = NSMakeSize(480, 122);
    
    [attackSlider sendActionOn:NSLeftMouseDraggedMask | NSLeftMouseDownMask];
    [releaseSlider sendActionOn:NSLeftMouseDraggedMask | NSLeftMouseDownMask];
}

-(void) updateAttack {
    attackField.stringValue = [attackParameter stringFromValue: nil];
    attackSlider.floatValue = (log10f(attackParameter.value) + 3.0) * 100.0f;
}

-(void) updateRelease {
    releaseField.stringValue = [releaseParameter stringFromValue: nil];
    releaseSlider.floatValue = (log10f(releaseParameter.value) + 3.0) * 100.0f;
}

-(IBAction) attackValueChanged:(id) sender {
    if (sender == attackField) {
        attackParameter.value = attackField.floatValue;
    } else if (sender == attackSlider) {
        attackParameter.value = powf(10.0, attackSlider.floatValue * 0.01 - 3.0);
    }
}

-(IBAction) releaseValueChanged:(id) sender {
    if (sender == releaseField) {
        releaseParameter.value = releaseField.floatValue;
    } else if (sender == releaseSlider) {
        releaseParameter.value = powf(10.0, releaseSlider.floatValue * 0.01 - 3.0);
    }
}

@end

File: Instrument/iOS/InstrumentDemoFramework/InstrumentDemoViewController.swift
/*
	Copyright (C) 2016 Apple Inc. All Rights Reserved.
	See LICENSE.txt for this sample’s licensing information
	
	Abstract:
	View controller for the InstrumentDemo audio unit. This is the app extension's principal class, responsible for creating both the audio unit and its view. Manages the interactions between a InstrumentView and the audio unit's parameters.
*/

import UIKit
import CoreAudioKit
import ActionKit

public class InstrumentDemoViewController: AUViewController { //, InstrumentViewDelegate {
    // MARK: Properties
	
    public var audioUnit: AUv3InstrumentDemo? {
      didSet {
        DispatchQueue.main.async {
          if self.isViewLoaded {
            self.connectViewWithAU()
          }
        }
      }
    }
  
	var parameterObserverToken: AUParameterObserverToken?
  let stack = UIStackView()
  var params: [AUParameterAddress: (AUParameter, ParameterView)] = [:]
  
  public override func loadView() {
    super.loadView()
    stack.axis = .vertical
    view.addSubview(stack)
    stack.translatesAutoresizingMaskIntoConstraints = false
    
    let constraints = [
      stack.topAnchor.constraint(equalTo: view.topAnchor),
      stack.leadingAnchor.constraint(equalTo: view.leadingAnchor),
      stack.trailingAnchor.constraint(equalTo: view.trailingAnchor)
    ]
    
    NSLayoutConstraint.activate(constraints)
  }
	
	public override func viewDidLoad() {
		super.viewDidLoad()
		
		// Respond to changes in the instrumentView (attack and/or release changes).
		
        guard audioUnit != nil else { return }

        connectViewWithAU()
	}
    
	/*
		We can't assume anything about whether the view or the AU is created first.
		This gets called when either is being created and the other has already 
        been created.
	*/
	func connectViewWithAU() {
		guard let paramTree = audioUnit?.parameterTree else { return }
    
    let groupStack = UIStackView()
    groupStack.axis = .vertical
    groupStack.spacing = 8.0
    stack.addArrangedSubview(groupStack)
    
    paramTree.children.forEach { group in
      guard let group = group as? AUParameterGroup else { return }
      group.allParameters.forEach { param in
        let paramView = ParameterView(param: param)
        
        groupStack.addArrangedSubview(paramView)
        paramView.slider.addControlEvent(.valueChanged) {
          param.value = paramView.slider.value
        }
        
        params[param.address] = (param, paramView)
        update(param: param, view: paramView)
      }
    }

		parameterObserverToken = paramTree.token(byAddingParameterObserver: { [weak self] address, value in
            guard let this = self, let uiParam = this.params[address] else { return }
			DispatchQueue.main.async {
        this.update(param: uiParam.0, view: uiParam.1)
			}
		})
  }
  
  func update(param: AUParameter, view: ParameterView) {
    view.displayValue = param.value
  }
}
